crates/gpui/src/platform/windows/direct_write.rs

```diff
@@ -0,0 +1,1286 @@
+        let advance = [0.0f32];
+        let offset = [DWRITE_GLYPH_OFFSET::default()];
+        let glyph_run = DWRITE_GLYPH_RUN {
+            fontFace: unsafe { std::mem::transmute_copy(&font.font_face) },
+            fontEmSize: params.font_size.0,
+            glyphCount: 1,
+            glyphIndices: glyph_id.as_ptr(),
+            glyphAdvances: advance.as_ptr(),
+            glyphOffsets: offset.as_ptr(),
+            isSideways: BOOL(0),
+            bidiLevel: 0,
+        };
+        let transform = DWRITE_MATRIX {
+            m11: params.scale_factor,
+            m12: 0.0,
+            m21: 0.0,
+            m22: params.scale_factor,
+            dx: 0.0,
+            dy: 0.0,
+        };
+        self.components.factory.CreateGlyphRunAnalysis(
+            &glyph_run as _,
+            1.0,
+            Some(&transform as _),
+            DWRITE_RENDERING_MODE_NATURAL,
+            DWRITE_MEASURING_MODE_NATURAL,
+            0.0,
+            0.0,
+        )
+    }
+
+    fn raster_bounds(&self, params: &RenderGlyphParams) -> Result<Bounds<DevicePixels>> {
+        unsafe {
+            let glyph_run_analysis = self.get_glyphrun_analysis(params)?;
+            let bounds = glyph_run_analysis.GetAlphaTextureBounds(DWRITE_TEXTURE_CLEARTYPE_3x1)?;
+
+            Ok(Bounds {
+                origin: Point {
+                    x: DevicePixels(bounds.left),
+                    y: DevicePixels(bounds.top),
+                },
+                size: Size {
+                    width: DevicePixels(bounds.right - bounds.left),
+                    height: DevicePixels(bounds.bottom - bounds.top),
+                },
+            })
+        }
+    }
+
+    fn glyph_for_char(&self, font_id: FontId, ch: char) -> Option<GlyphId> {
+        let font_info = &self.fonts[font_id.0];
+        let codepoints = [ch as u32];
+        let mut glyph_indices = vec![0u16; 1];
+        unsafe {
+            font_info
+                .font_face
+                .GetGlyphIndices(codepoints.as_ptr(), 1, glyph_indices.as_mut_ptr())
+                .log_err()
+        }
+        .map(|_| GlyphId(glyph_indices[0] as u32))
+    }
+
+    fn rasterize_glyph(
+        &self,
+        params: &RenderGlyphParams,
+        glyph_bounds: Bounds<DevicePixels>,
+    ) -> Result<(Size<DevicePixels>, Vec<u8>)> {
+        if glyph_bounds.size.width.0 == 0 || glyph_bounds.size.height.0 == 0 {
+            return Err(anyhow!("glyph bounds are empty"));
+        }
+
+        let font_info = &self.fonts[params.font_id.0];
+        let glyph_id = [params.glyph_id.0 as u16];
+        let advance = [glyph_bounds.size.width.0 as f32];
+        let offset = [DWRITE_GLYPH_OFFSET {
+            advanceOffset: -glyph_bounds.origin.x.0 as f32 / params.scale_factor,
+            ascenderOffset: glyph_bounds.origin.y.0 as f32 / params.scale_factor,
+        }];
+        let glyph_run = DWRITE_GLYPH_RUN {
+            fontFace: unsafe { std::mem::transmute_copy(&font_info.font_face) },
```

Comment:
Same for here

Reply:
In fact, this is the recommended practice to do so.